File: leetCode/easy/1528-shuffleString.ts
import { assertEquals } from "https://deno.land/std@0.182.0/testing/asserts.ts";
import { describe, it } from "https://deno.land/std@0.182.0/testing/bdd.ts";

function shuffleString(s: string, indices: number[]) {
  return indices
    .reduce((newStr, indice, idx) => {
      newStr[indice] = s[idx];

      return newStr;
    }, Array.from({ length: s.length }))
    .join("");
}

export { shuffleString };

describe("shuffleString", () => {
  it("1. should return a string with the characters at the indices shuffled", () => {
    const s = "codeleet";
    const indices = [4, 5, 6, 7, 0, 2, 1, 3];
    const result = "leetcode";
    assertEquals(shuffleString(s, indices), result);
  });

  it("2. should return a string with the characters at the indices shuffled", () => {
    const s = "abc";
    const indices = [0, 1, 2];
    const result = "abc";
    assertEquals(shuffleString(s, indices), result);
  });

  it("3. should return a string with the characters at the indices shuffled", () => {
    const s = "aiohn";
    const indices = [3, 1, 4, 2, 0];
    const result = "nihao";
    assertEquals(shuffleString(s, indices), result);
  });
});

File: leetCode/easy/1528-shuffleString.jl
function shuffleString(s::String, indices::Array{Int64,1})
  newStr = Array{String,1}(undef, length(s))

  for (idx, indice) in enumerate(indices)
    newStr[indice] = string(s[idx])
  end

  return join(newStr)
end

export shuffleString

using Test

Test.@testset "shuffleString" begin
  Test.@test shuffleString("codeleet", [5, 6, 7, 8, 1, 3, 2, 4]) == "leetcode"
  Test.@test shuffleString("abc", [1, 2, 3]) == "abc"
  Test.@test shuffleString("aiohn", [4, 2, 5, 3, 1]) == "nihao"
  Test.@test shuffleString("aaiougrt", [5, 1, 3, 7, 8, 4, 2, 6]) == "arigatou"
  Test.@test shuffleString("art", [2, 1, 3]) == "rat"
end